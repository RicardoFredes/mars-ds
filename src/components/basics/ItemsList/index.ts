/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 */

import ItemsList from "./items-list.component";
export default ItemsList;
export * from "./items-list.types";
