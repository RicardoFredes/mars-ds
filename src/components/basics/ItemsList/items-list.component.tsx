import type { ItemsListProps } from "./items-list.types";

import classNames from "classnames";

import Icon from "@/components/basics/Icon";
import Link from "@/components/basics/Link";
import Subtitle from "@/components/typographies/Subtitle";
import Text from "@/components/typographies/Text";

const ItemsList = ({ list, className, componentLink, ...props }: ItemsListProps) => {
  const cn = classNames("items-list", className);
  const ComponentLink = componentLink || Link;

  return (
    <ol className={cn} {...props}>
      {list.map(
        (
          {
            title,
            subtitle,
            link,
            icon,
            iconColor = "var(--color-background-primary)",
            circleColor = "var(--color-text-primary)",
            children,
          },
          index
        ) => {
          const itemCn = classNames("items-list__item", {
            "items-list__item--has-icon": icon,
          });

          const containerCn = classNames("items-list__container", {
            "items-list__container--has-icon": icon,
          });

          const circleCn = classNames("items-list__circle", {
            "items-list__circle--has-icon": icon,
          });

          return (
            <li className={itemCn} key={index}>
              <ComponentLink {...link} className="items-list__link">
                <div className={containerCn}>
                  <div
                    className={circleCn}
                    data-testid="items-list__circle"
                    style={{ backgroundColor: circleColor }}
                  >
                    {icon && (
                      <Icon
                        name={icon}
                        className="items-list__icon"
                        data-testid="items-list__icon"
                        color={iconColor}
                      />
                    )}
                  </div>

                  <div className="items-list__content">
                    <Text className="items-list__title" data-testid="items-list__title">
                      {title}
                    </Text>
                    {subtitle && (
                      <Subtitle className="items-list__subtitle" data-testid="items-list__subtitle">
                        {subtitle}
                      </Subtitle>
                    )}
                  </div>
                </div>
              </ComponentLink>

              {children && (
                <div className="items-list__children" data-testid="items-list__children">
                  {children}
                </div>
              )}
            </li>
          );
        }
      )}
    </ol>
  );
};

export default ItemsList;
