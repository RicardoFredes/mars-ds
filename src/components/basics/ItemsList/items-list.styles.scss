.items-list {
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: 0;
  list-style-type: none;

  &__link,
  .link {
    width: 100%;
    color: $color-text-primary;
    text-decoration: none;

    &:visited,
    &:hover {
      color: $color-text-primary;
      text-decoration: none;
    }
  }

  &__item {
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    &:not(:last-child) {
      &::after {
        position: absolute;
        top: 56px;
        left: 20px;
        width: 1px;
        height: calc(100% - 40px);
        content: "";
        border-left: 1px solid $color-divider;
      }
    }

    &--has-icon {
      min-height: 80px;

      &:not(:last-child) {
        &::after {
          top: 64px;
          left: 28px;
          height: calc(100% - 56px);
        }
      }
    }
  }

  &__container {
    display: flex;
    gap: 0 $gap-sm-4;
    align-items: center;
    justify-content: flex-start;
    width: 100%;
    min-height: 72px;
    padding: 0 $spacing-md;
    cursor: pointer;
    transition: background-color ease 0.2s;

    &:focus,
    &:hover {
      background-color: $color-neutral-ghost;
    }
  }

  &__circle {
    $icon-size: 16px;

    display: flex;
    align-items: center;
    justify-content: center;
    min-width: $icon-size;
    min-height: $icon-size;
    border-radius: 50%;

    &--has-icon {
      $icon-size: 32px;

      min-width: $icon-size;
      min-height: $icon-size;
    }
  }

  &__icon {
    > * {
      padding: 2px;
    }
  }

  &__subtitle {
    color: $color-neutral-700;
  }

  &__content {
    width: 100%;
  }

  &__children {
    width: calc(100% - calc($spacing-xl * 2));
    margin: 0 0 0 calc($spacing-xl * 2);
  }
}
