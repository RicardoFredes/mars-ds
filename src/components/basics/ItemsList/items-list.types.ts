import type { HTMLProps, ReactNode } from "react";

import type { LinkProps } from "@/components/basics/Link";
import type { ComponentLink } from "@/types";

interface ItemsListElementProps extends HTMLProps<HTMLLIElement> {
  title: string;
  link?: LinkProps;
  subtitle?: string;
  children?: ReactNode;
  icon?: string;
  iconColor?: string;
  circleColor?: string;
}

interface ItemsListProps extends Omit<HTMLProps<HTMLOListElement>, "list" | "type"> {
  componentLink?: ComponentLink;
  list: ItemsListElementProps[];
}

export type { ItemsListProps, ItemsListElementProps };
