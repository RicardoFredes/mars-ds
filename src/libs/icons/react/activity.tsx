/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconActivity = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12" />
</svg>
);
  
export default IconActivity;