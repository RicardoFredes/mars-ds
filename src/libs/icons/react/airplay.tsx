/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAirplay = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  stroke-linejoin="round"
>
  <path d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1" />
  <polygon points="12 15 17 21 7 21 12 15" />
</svg>
);
  
export default IconAirplay;