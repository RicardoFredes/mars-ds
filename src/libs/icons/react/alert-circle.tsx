/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAlertCircle = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <circle cx="12" cy="12" r="10" />
  <line x1="12" y1="8" x2="12" y2="12" />
  <line x1="12" y1="16" x2="12.01" y2="16" />
</svg>
);
  
export default IconAlertCircle;