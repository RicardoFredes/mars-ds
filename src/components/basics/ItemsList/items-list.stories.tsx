import type { ItemsListProps } from "./items-list.types";
import type { ComponentMeta, ComponentStory } from "@storybook/react";

import ItemsList from "./items-list.component";

export default {
  title: "Components/ItemsList",
  component: ItemsList,
  argTypes: {},
} as ComponentMeta<typeof ItemsList>;

const Template: ComponentStory<typeof ItemsList> = (args) => <ItemsList {...args} />;

const DefaultArgs: ItemsListProps = {
  list: [
    {
      title: "A soma de forma intuitiva",
      subtitle: "8 minutos",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "video",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
    {
      title: "Exercícios de compreensão sobre soma",
      subtitle: "3 exercícios",
      children: (
        <ItemsList
          list={[
            {
              title: "Exercício 1",
              link: {
                href: "https://mesalva.com",
                target: "_blank",
              },
              circleColor: "var(--color-text-disabled)",
            },
            {
              title: "Exercício 2",
              link: {
                href: "https://mesalva.com",
                target: "_blank",
              },
              circleColor: "var(--color-text-disabled)",
            },
            {
              title: "Exercício 3",
              link: {
                href: "https://mesalva.com",
                target: "_blank",
              },
              circleColor: "var(--color-text-disabled)",
            },
          ]}
        />
      ),
      icon: "edit",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
    {
      title: "Somando de tudo",
      subtitle: "10 minutos",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "video",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
    {
      title: "Exercícios de compreensão sobre soma II",
      subtitle: "3 exercícios",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "edit",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
    {
      title: "Somando de tudo",
      subtitle: "9 minutos",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "video",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
    {
      title: "Exercícios de compreensão sobre subtração",
      subtitle: "3 exercícios",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "edit",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
    {
      title: "Exercícios de compreensão sobre subtração de naturais",
      subtitle: "5 exercícios",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "edit",
      iconColor: "var(--color-background-primary)",
      circleColor: "var(--color-text-primary)",
    },
  ],
};

export const Default = Template.bind({});
Default.args = { ...DefaultArgs };
