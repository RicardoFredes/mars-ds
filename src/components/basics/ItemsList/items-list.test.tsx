import type { ItemsListProps, ItemsListElementProps } from "./items-list.types";

import { render } from "@testing-library/react";

import ItemsList from "./items-list.component";

const CLASSES = {
  item: "items-list__item",
  container: "items-list__container",
  circle: "items-list__circle",
  icon: "items-list__icon",
  content: "items-list__content",
  title: "items-list__title",
  subtitle: "items-list__subtitle",
  children: "items-list__children",
};

const defaultProps: ItemsListProps = {
  list: [
    {
      title: "A soma de forma intuitiva",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
    },
    {
      title: "Somando de tudo",
      subtitle: "10 minutos",
      link: {
        href: "https://mesalva.com",
        target: "_blank",
      },
      icon: "video",
      iconColor: "var(--color-white)",
      circleColor: "var(--color-text-primary)",
    },
  ],
};

const makeItemProps = (overrides: Partial<ItemsListElementProps> = {}): ItemsListElementProps => ({
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  ...defaultProps.list[0]!,
  ...overrides,
});

const makeSut = (props?: ItemsListProps) => render(<ItemsList {...defaultProps} {...props} />);

describe("<ItemsList>", () => {
  it("should match the snapshot", () => {
    const wrapper = makeSut();
    expect(wrapper).toMatchSnapshot();
  });

  it("should contain a link wrapping each item", () => {
    const wrapper = makeSut();
    const links = wrapper.getAllByRole("link") as HTMLAnchorElement[];

    expect(links).toHaveLength(defaultProps.list.length);
    expect(
      links.forEach((link) => {
        expect(link).toHaveAttribute("href");
        expect(link.attributes.getNamedItem("href")?.value).toBe(
          defaultProps.list?.[0]?.link?.href
        );

        expect(link).toHaveAttribute("target");
        expect(link.attributes.getNamedItem("target")?.value).toBe(
          defaultProps.list?.[0]?.link?.target
        );
      })
    );
  });

  describe("dots", () => {
    it("should contain a dot for each item", () => {
      const wrapper = makeSut();
      const dots = wrapper.getAllByTestId(CLASSES.circle);

      expect(dots.length).toBe(defaultProps.list.length);
    });

    it("should change color based on the item's circleColor prop", () => {
      const props: ItemsListProps = {
        list: [makeItemProps({ circleColor: "var(--color-text-disabled)" })],
      };

      const wrapper = makeSut(props);
      const dot = wrapper.getByTestId(CLASSES.circle);

      expect(dot).toHaveStyle(`background-color: ${props.list?.[0]?.circleColor}`);
    });
  });

  describe("when the icon prop is present", () => {
    it("should contain an icon", () => {
      const props: ItemsListProps = {
        list: [makeItemProps({ icon: "video" })],
      };
      const wrapper = makeSut(props);
      const icons = wrapper.getAllByTestId(CLASSES.icon);

      expect(icons.length).toBe(1);
    });

    it("should change color based on the item's iconColor prop", () => {
      const props: ItemsListProps = {
        list: [makeItemProps({ icon: "video", iconColor: "var(--color-text-disabled)" })],
      };

      const wrapper = makeSut(props);
      const icon = wrapper.getByTestId(CLASSES.icon);

      expect(icon).toHaveStyle(`background-color: ${props.list?.[0]?.circleColor}`);
    });
  });

  it("should contain a title", () => {
    const wrapper = makeSut();
    const titles = wrapper.getAllByTestId(CLASSES.title);

    expect(titles.length).toBe(defaultProps.list.length);
  });

  it("should contain a subtitle", () => {
    const props: ItemsListProps = {
      list: [makeItemProps({ subtitle: "8 minutos" })],
    };
    const wrapper = makeSut(props);
    const subtitles = wrapper.getAllByTestId(CLASSES.subtitle);

    expect(subtitles.length).toBe(1);
  });

  it("should contain children", () => {
    const props: ItemsListProps = {
      list: [makeItemProps({ children: <p>children</p> })],
    };
    const wrapper = makeSut(props);
    const children = wrapper.getAllByTestId(CLASSES.children);

    expect(children.length).toBe(1);
  });
});
