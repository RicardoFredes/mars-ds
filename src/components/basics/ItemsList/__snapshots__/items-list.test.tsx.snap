// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemsList> should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ol
        class="items-list"
      >
        <li
          class="items-list__item"
        >
          <a
            class="link items-list__link link--primary"
            href="https://mesalva.com"
            target="_blank"
          >
             
            <div
              class="items-list__container"
            >
              <div
                class="items-list__circle"
                data-testid="items-list__circle"
              />
              <div
                class="items-list__content"
              >
                <p
                  class="text text--size-md items-list__title"
                  data-testid="items-list__title"
                >
                  A soma de forma intuitiva
                </p>
              </div>
            </div>
          </a>
        </li>
        <li
          class="items-list__item items-list__item--has-icon"
        >
          <a
            class="link items-list__link link--primary"
            href="https://mesalva.com"
            target="_blank"
          >
             
            <div
              class="items-list__container items-list__container--has-icon"
            >
              <div
                class="items-list__circle items-list__circle--has-icon"
                data-testid="items-list__circle"
              >
                <span
                  class="icon items-list__icon"
                  data-name="video"
                  data-testid="items-list__icon"
                >
                  <svg
                    fill="none"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M21 7.15C20.3789 6.87035 19.6509 6.98841 19.15 7.45L17 9.45V8C17 6.34315 15.6569 5 14 5H5C3.34315 5 2 6.34315 2 8V16C2 17.6569 3.34315 19 5 19H14C15.6569 19 17 17.6569 17 16V14.55L19.16 16.55C19.478 16.8378 19.8911 16.9981 20.32 17C20.558 16.9994 20.7932 16.9483 21.01 16.85C21.6133 16.606 22.0087 16.0208 22.01 15.37V8.63C22.007 7.97674 21.6072 7.39089 21 7.15Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="items-list__content"
              >
                <p
                  class="text text--size-md items-list__title"
                  data-testid="items-list__title"
                >
                  Somando de tudo
                </p>
                <h3
                  class="subtitle subtitle--size-md items-list__subtitle"
                  data-testid="items-list__subtitle"
                >
                  10 minutos
                </h3>
              </div>
            </div>
          </a>
        </li>
      </ol>
    </div>
  </body>,
  "container": <div>
    <ol
      class="items-list"
    >
      <li
        class="items-list__item"
      >
        <a
          class="link items-list__link link--primary"
          href="https://mesalva.com"
          target="_blank"
        >
           
          <div
            class="items-list__container"
          >
            <div
              class="items-list__circle"
              data-testid="items-list__circle"
            />
            <div
              class="items-list__content"
            >
              <p
                class="text text--size-md items-list__title"
                data-testid="items-list__title"
              >
                A soma de forma intuitiva
              </p>
            </div>
          </div>
        </a>
      </li>
      <li
        class="items-list__item items-list__item--has-icon"
      >
        <a
          class="link items-list__link link--primary"
          href="https://mesalva.com"
          target="_blank"
        >
           
          <div
            class="items-list__container items-list__container--has-icon"
          >
            <div
              class="items-list__circle items-list__circle--has-icon"
              data-testid="items-list__circle"
            >
              <span
                class="icon items-list__icon"
                data-name="video"
                data-testid="items-list__icon"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21 7.15C20.3789 6.87035 19.6509 6.98841 19.15 7.45L17 9.45V8C17 6.34315 15.6569 5 14 5H5C3.34315 5 2 6.34315 2 8V16C2 17.6569 3.34315 19 5 19H14C15.6569 19 17 17.6569 17 16V14.55L19.16 16.55C19.478 16.8378 19.8911 16.9981 20.32 17C20.558 16.9994 20.7932 16.9483 21.01 16.85C21.6133 16.606 22.0087 16.0208 22.01 15.37V8.63C22.007 7.97674 21.6072 7.39089 21 7.15Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </div>
            <div
              class="items-list__content"
            >
              <p
                class="text text--size-md items-list__title"
                data-testid="items-list__title"
              >
                Somando de tudo
              </p>
              <h3
                class="subtitle subtitle--size-md items-list__subtitle"
                data-testid="items-list__subtitle"
              >
                10 minutos
              </h3>
            </div>
          </div>
        </a>
      </li>
    </ol>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
